src/compiler/checker.ts

```diff
@@ -9514,100 +9674,140 @@ module ts {
+                    typeAsIterator.iteratorElementType = (<GenericType>iterator).typeArguments[0];
                 }
+                else {
+                    let iteratorNextFunction = getTypeOfPropertyOfType(iterator, "next");
+                    if (iteratorNextFunction && allConstituentTypesHaveKind(iteratorNextFunction, TypeFlags.Any)) {
+                        return undefined;
+                    }
 
-                let iteratorNextFunctionSignatures = iteratorNextFunction ? getSignaturesOfType(iteratorNextFunction, SignatureKind.Call) : emptyArray;
-                if (iteratorNextFunctionSignatures.length === 0) {
-                    if (errorNode) {
-                        error(errorNode, Diagnostics.An_iterator_must_have_a_next_method);
+                    let iteratorNextFunctionSignatures = iteratorNextFunction ? getSignaturesOfType(iteratorNextFunction, SignatureKind.Call) : emptyArray;
+                    if (iteratorNextFunctionSignatures.length === 0) {
+                        if (errorNode) {
+                            error(errorNode, Diagnostics.An_iterator_must_have_a_next_method);
+                        }
+                        return undefined;
                     }
-                    return undefined;
-                }
 
-                let iteratorNextResult = getUnionType(map(iteratorNextFunctionSignatures, getReturnTypeOfSignature));
-                if (allConstituentTypesHaveKind(iteratorNextResult, TypeFlags.Any)) {
-                    return undefined;
-                }
+                    let iteratorNextResult = getUnionType(map(iteratorNextFunctionSignatures, getReturnTypeOfSignature));
+                    if (allConstituentTypesHaveKind(iteratorNextResult, TypeFlags.Any)) {
+                        return undefined;
+                    }
 
-                let iteratorNextValue = getTypeOfPropertyOfType(iteratorNextResult, "value");
-                if (!iteratorNextValue) {
-                    if (errorNode) {
-                        error(errorNode, Diagnostics.The_type_returned_by_the_next_method_of_an_iterator_must_have_a_value_property);
+                    let iteratorNextValue = getTypeOfPropertyOfType(iteratorNextResult, "value");
+                    if (!iteratorNextValue) {
+                        if (errorNode) {
+                            error(errorNode, Diagnostics.The_type_returned_by_the_next_method_of_an_iterator_must_have_a_value_property);
+                        }
+                        return undefined;
                     }
-                    return undefined;
+
+                    typeAsIterator.iteratorElementType = iteratorNextValue;
                 }
+            }
+
+            return typeAsIterator.iteratorElementType;
+        }
 
-                return iteratorNextValue;
+        function getElementTypeFromIterableIterator(iterableIterator: Type): Type {

```

Comment:
Weird how `getElementTypeFromIterable` and `getElementTypeFromIterator` both take an `errorNode` but you don't here.